cedar-lean/Cedar/Spec/PartialExpr.lean

```diff
@@ -0,0 +1,212 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data
+import Cedar.Spec.Expr
+import Cedar.Thm.Data.Map -- necessary for a termination argument
+import Cedar.Thm.Termination.Expr
+import Cedar.Thm.Termination.Value
+
+/-! This file defines abstract syntax for Cedar expressions. -/
+
+namespace Cedar.Spec
+
+open Cedar.Data
+
+/--
+  Identical to `Expr` except that it has an `unknown` case, and the recursive
+  elements are also all `PartialExpr` instead of `Expr`
+-/
+inductive PartialExpr where
+  | lit (p : Prim)
+  | var (v : Var)
+  | ite (cond : PartialExpr) (thenExpr : PartialExpr) (elseExpr : PartialExpr)
+  | and (a : PartialExpr) (b : PartialExpr)
+  | or (a : PartialExpr) (b : PartialExpr)
+  | unaryApp (op : UnaryOp) (expr : PartialExpr)
+  | binaryApp (op : BinaryOp) (a : PartialExpr) (b : PartialExpr)
+  | getAttr (expr : PartialExpr) (attr : Attr)
+  | hasAttr (expr : PartialExpr) (attr : Attr)
+  | set (ls : List PartialExpr)
+  | record (map : List (Attr × PartialExpr))
+  | call (xfn : ExtFun) (args : List PartialExpr)
+  | unknown (name : String)
```

Comment:
I'd recommend defining an abbreviation for unknown and using that everywhere instead of `String`, just in case we change our mind about how to model unknowns.

```
abbrev Unknown := String

...

  |  unknown (u : Unknown)
```